src/app/shared/utils/deduplicate-name.ts

```diff
@@ -0,0 +1,25 @@
+import { normalizeName } from './normalize-name';
+
+/**
+ * Deduplicate a name by adding a number to the end if it already exists.
+ */
+export function deduplicateName(name: string, existingNames: Array<string>): string {
+  const normalizedNames = existingNames.map((n) => normalizeName(n));
+
+  // Check if name already exists
+  while (normalizedNames.includes(normalizeName(name))) {
+    // Check if the name already has a number
+    const lastNumber = name.match(/\d+$/);
+    if (lastNumber) {
+      // Increment the number
+      const number = parseInt(lastNumber[0], 10);
+      name = name.replace(/\d+$/, '') + (number + 1);
+    } else {
+      // Add a number to the color name
+      name += ' 2';
+    }
+  }
```

Comment:
@kevinknecht-ai Do you have a better idea than this?

Reply:
Maybe if we could "calculate" the correct ending number instead of guessing it `n` times. 

For this we could do something like:

1. Find all existing names that are of the form ${newName}\\d* (new color name + maybe a number)
2. Get all the numbers that already exist
3. Take the lowest non-existing number bigger than 2 or a number bigger by one than all of the already existing numbers 


The following code expands on this idea HOWEVER THIS CODE IS UNTESTED so it should be handled more like pseudo code^^ 

``` ts

/**
 * Returns a unique name by attaching a number the end of the name if it already exists 
 */
export function deduplicateName(name: string, existingNames: Array<string>): string {
  const normalizedNames = existingNames.map((n) => normalizeName(n));

  if(!normalizedNames.includes(name)) {
    return name; 
  }

  // that means we first find all existing names that start with the new name and have a number at the end
  const namingPattern = new RegExp(`${name}\\d*$`)
  const similarPrefix = normalizedNames.filter((existingName) => namingPattern.test(existingName))

  // if no strings are similar (meaning equal or have this name as prefix) return the input
  if(!similarPrefix) {
    return name;
  }

  // the next step is to get all the ending numbers
  // typescript gives me type (number | null)[] but actually I want number[] (don't know if this is allowed)
  // however "null" should be the case where the color name doesn't have a number at the end (this could be mapped to 1 though). 
  // This doesn't have to be handled because we check if(!normalizedNames.includes(name))
  const endingNumbers = similarPrefix.map((existingName) => getLastNumberOrNull(existingName)).filter(Boolean) as number[];

  // unfortunately we can't check if the highest number of the array is equal to the length of the array because we can't assume that the ending numbers in existingNames are optimal (starting at the number 2 counting up to the number of same "names")
  // otherwise we could concat max(endingNumbers) + 1  to the input name

  // find the lowest number that does not exist in the array OR the highest found + 1 as a unique suffix

  const uniqueAndSortedNumbers = [... new Set(endingNumbers)].sort()

  // if all naming was done correctly we have an array like [2,..., similarPrefix.length+1] as the naming should start with 2 and should increase
  // that means the actual value in the array should be equal to the index + 1
  const uniqueLowestNumber = findLowestMissingPositive(uniqueAndSortedNumbers);

  // increase the number 
  return name + (uniqueLowestNumber + 1);
}

/**
 * Just an ai generated function that finds the lowest missing positive number in this array
 * No guarantee at this point that this function delivers the correct output^^
 */
function findLowestMissingPositive(nums: number[]): number {
  if (nums.length === 0 || nums[0] > 2) {
    return 2;
  }

  for (let i = 0; i < nums.length; i++) {
    if (nums[i] !== i + 2) {
      return i + 2;
    }
  }

  return nums.length + 2;
}

/**
 * returns the last number of a given string if exits 
 */
function getLastNumberOrNull(name: string): number | null {
    // Check if the name already has a number
    const lastNumber = name.match(/\d+$/);
    if(lastNumber) {
      return parseInt(lastNumber[0]);
    }
    return null;
}

```

Reply:
I thought about it again and even if your suggestion looks good, let's leave it as is. It's just an edge case with a very low change of occurring and my primitive solution should be fine, even if maybe not performant for higher numbers (that will likely never happen).

Reply:
Yes, of course. The option was just out of curiousity and not necessity (therefore diregarding use/cost)